lib/runner/extensions/waterfall.command.js

```diff
@@ -158,14 +166,22 @@ module.exports = {
 
                 nextCoords = _.clone(coords);
 
-                // if the position was detected, we set the position to the one previous to the desired location
-                // this ensures that the next call to .whatnext() will return the desired position.
-                snr.valid ? (nextCoords.position = snr.position - 1) :
-                    // if snr was requested, but not valid, we stop this iteration.
-                    // stopping an iteration is equivalent to seeking the last position of the current
-                    // iteration, so that the next call to .whatnext() will automatically move to the next
-                    // iteration.
-                    ((snr.defined || executionError) && (nextCoords.position = nextCoords.length - 1));
+                if (snr.valid) {
+                    nextCoords.position = snr.position - 1;
+                }
+                else {
+                    (snr.defined || executionError) && (nextCoords.position = nextCoords.length - 1);
+                    (stopOnFailure && executionError) && (stopRunNow = true);
+                }
+
+                // // if the position was detected, we set the position to the one previous to the desired location
```

Comment:
delete this commented block, retain the comment between if-else and before if. no harm.